src/core/thread/mlr_manager.hpp

```diff
@@ -0,0 +1,148 @@
+
+/**
+ * @addtogroup core-mlr
+ *
+ * @brief
+ *   This module includes definitions for Multicast Listener Registration.
+ *
+ * @{
+ *
+ * @defgroup core-mlr Mlr
+ *
+ * @}
+ *
+ */
+
+/**
+ * This class implements MLR management.
+ *
+ */
+class MlrManager : public InstanceLocator, public Notifier::Receiver
+{
+public:
+    /**
+     * This constructor initializes the object.
+     *
+     * @param[in]  aInstance     A reference to the OpenThread instance.
+     *
+     */
+    explicit MlrManager(Instance &aInstance);
+
+    /**
+     * This method notifies Primary Backbone Router status.
+     *
+     * @param[in]  aState   The state or state change of Primary Backbone Router.
+     * @param[in]  aConfig  The Primary Backbone Router service.
+     *
+     */
+    void UpdateBackboneRouterPrimary(BackboneRouter::Leader::State               aState,
+                                     const BackboneRouter::BackboneRouterConfig &aConfig);
+
+    /**
+     * This method returns the Multicast Listener Table.
+     *
+     * @returns The Multicast Listener Table.
+     */
+    MulticastListenerTable &GetMulticastListenerTable() { return mMulticastListenerTable; }
+
+    /**
+     * This method notifies MLR reregistration or renewing to start.
+     *
+     */
+    void Reregistration();
+
+private:
+    enum
+    {
+        kTimerInterval = 1000,
+    };
+
+    static void HandleNotifierEvents(Notifier::Receiver &aReceiver, Events aEvents);
+    void        HandleNotifierEvents(Events aEvents);
+
+    void UpdateLocalSubscribtions();
+
+    void SendMulticastListenerRegistration();
+
+    static void HandleMulticastListenerRegistrationResponse(void *               aContext,
+                                                            otMessage *          aMessage,
+                                                            const otMessageInfo *aMessageInfo,
+                                                            otError              aResult)
+    {
+        static_cast<MlrManager *>(aContext)->HandleMulticastListenerRegistrationResponse(
+            static_cast<Coap::Message *>(aMessage), static_cast<const Ip6::MessageInfo *>(aMessageInfo), aResult);
+    }
+
+    void HandleMulticastListenerRegistrationResponse(Coap::Message *         aMessage,
+                                                     const Ip6::MessageInfo *aMessageInfo,
+                                                     otError                 aResult);
+
+    void ScheduleSend(uint16_t aDelay);
+    void ResetTimer(void);
+    void UpdateReregistrationPeriod(bool aRereg);
+
+    static void HandleTimer(Timer &aTimer) { aTimer.GetOwner<MlrManager>().HandleTimer(); }
+    void        HandleTimer(void);
+
+    void LogMulticastListenerTable();
+
+    MulticastListenerTable mMulticastListenerTable;
+    TimerMilli             mTimer;
+    uint32_t               mReregistrationDelay;
+    uint16_t               mSendDelay;
+    bool                   mCoapPending : 1;
```

Comment:
How about renaming `mCoapPending ` as `mMlrPending` or `mPending` or `mWaitingforResponse` or `mMlrInProgress`?

Reply:
I used to call `Reregistration` from `MleRouter::HandleStateTimer`. But it's not so now.

Reply:
renamed to `mMlrPending`